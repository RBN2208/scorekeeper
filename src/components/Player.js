import styled from 'styled-components/macro'
import Button from './Button'

const MAX_HUE_VALUE = 120 // 120 is green
const MIN_HUE_VALUE = 0 // 0 is red

export default function Player({ name, score, onMinus, onPlus }) {
  const maxHue = Math.min(score * 10, MAX_HUE_VALUE)
  const hue = Math.max(maxHue, MIN_HUE_VALUE)
  const color = score === 0 ? 'black' : `hsl(${hue}, 50%, 50%)`
  return (
    <PlayerWrapper>
      {name}
      <PlayerScore color={color}>
        <ButtonStyled onClick={onMinus}>-</ButtonStyled>
        <span>{score}</span>
        <ButtonStyled onClick={onPlus}>+</ButtonStyled>
      </PlayerScore>
    </PlayerWrapper>
  )
}

const PlayerWrapper = styled.section`
  display: flex;
  align-items: center;
  justify-content: space-between;
`
const PlayerScore = styled.div`
  color: ${props => props.color || 'black'};
  display: grid;
  gap: 5px;
  grid-template-columns: repeat(3, 1fr);
  place-items: center;
  span {
    filter: drop-shadow(0 0 2px white);
  }
`

const ButtonStyled = styled(Button)`
  width: 32px;
  height: 32px;
  color: #373737;
`
