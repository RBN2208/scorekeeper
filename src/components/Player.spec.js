import { render, screen } from '@testing-library/react'
import userEvent from '@testing-library/user-event'
import Player from './Player'

describe('Player', () => {
  it('has two buttons', () => {
    render(<Player />)
    expect(screen.getAllByRole('button', /[+-]/)).toHaveLength(2)
  })
  it('can use buttons with plus and minus', () => {
    const callBackMinus = jest.fn()
    const callBackPlus = jest.fn()
    render(<Player onMinus={callBackMinus} onPlus={callBackPlus} />)
    userEvent.click(screen.getByRole('button', { name: '+' }))
    userEvent.click(screen.getByRole('button', { name: '-' }))
    expect(callBackPlus).toHaveBeenCalledTimes(1)
    expect(callBackMinus).toHaveBeenCalledTimes(1)
  })
  it('changes color on click', () => {
    const { rerender } = render(<Player name="John" score={0} />)
    expect(screen.getByText('0')).toMatchSnapshot()
    rerender(<Player name="John" score={10} />)
    expect(screen.getByText('10')).toMatchSnapshot()
    rerender(<Player name="John" score={-10} />)
    expect(screen.getByText('-10')).toMatchSnapshot()
  })
})
