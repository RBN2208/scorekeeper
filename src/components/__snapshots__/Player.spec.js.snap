// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Player changes color on click 1`] = `
<span>
  0
</span>
`;

exports[`Player changes color on click 2`] = `
<span>
  10
</span>
`;

exports[`Player changes color on click 3`] = `
<span>
  -10
</span>
`;
